Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp

```diff
@@ -41,22 +43,41 @@ UnifiedTextReplacementController::UnifiedTextReplacementController(WebPage& webP
 {
 }
 
-void UnifiedTextReplacementController::didBeginTextReplacementSession(const WTF::UUID& uuid)
+void UnifiedTextReplacementController::willBeginTextReplacementSession(const WTF::UUID& uuid, CompletionHandler<void(const Vector<WebUnifiedTextReplacementContextData>&)>&& completionHandler)
 {
-    RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::didBeginTextReplacementSession (%s)", uuid.toString().utf8().data());
+    RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::willBeginTextReplacementSession (%s)", uuid.toString().utf8().data());
 
     if (!m_webPage)
         return;
 
-    auto contextRange = m_webPage->rangeSelectionContext();
+    auto* corePage = m_webPage->corePage();
+    if (!corePage)
+        return;
+
+    Ref frame = CheckedRef(corePage->focusController())->focusedOrMainFrame();
+
+    auto contextRange = m_webPage->autocorrectionContextRange();
     if (!contextRange)
         return;
 
     auto liveRange = createLiveRange(*contextRange);
 
-    ASSERT(!m_sessionRanges.contains(uuid));
+    ASSERT(!m_contextRanges.contains(uuid));
+
+    m_contextRanges.set(uuid, liveRange);
+
+    auto selectedTextRange = frame->selection().selection().firstRange();
 
-    m_sessionRanges.set(uuid, liveRange);
+    auto attributedStringFromRange = attributedString(*contextRange);
+
+    auto selectedTextCharacterRange = WebCore::characterRange(*contextRange, *selectedTextRange);
+
+    completionHandler({ WebUnifiedTextReplacementContextData { WTF::UUID { 0 }, attributedStringFromRange, selectedTextCharacterRange } });
```

Comment:
Optional nit / Stylistic remark: Lots of blank lines in this method.

Reply:
I like blank lines, makes things easier to read personally